### ISSUE TITLE
Q.commutative contradicts is_commutative

### ISSUE THREAD

**Issue by eric-wieser** [sympy/sympy#19308]
I'm confused by this behavior:
```python
>>> from sympy import *
>>> x = Mul(Integer(-1), Determinant(MatrixSymbol(Symbol('f'), Integer(2), Integer(2))))
>>> x.is_commutative
False
>>> ask(Q.commutative(x))
True
```

Is there a bug here? Or do the two "commutative"s mean very different things?

**Comment by oscarbenjamin** [comment]
This is very strange:
```julia
In [1]: M = MatrixSymbol('M', 2, 2)                                                                                               

In [2]: e1 = Mul(-1, Determinant(M))                                                                                              

In [3]: e2 = - Determinant(M)                                                                                                     

In [4]: e1.is_commutative                                                                                                         
Out[4]: False

In [5]: e2.is_commutative                                                                                                         

In [6]: e1 == e2                                                                                                                  
Out[6]: True
```

**Comment by asmeurer** [comment]
The is_commutative is clearly wrong. Determinant is a number, so it is commutative. There is probably some incorrect logic somewhere that assumes that since the arguments are noncommutative, the expression is.

**Comment by eric-wieser** [comment]
I assume matrices of non-commutative symbols aren't supported anyway at the moment?

**Comment by asmeurer** [comment]
They might work with Matrix, though probably by accident. I don't know if it is tested. MatrixSymbol definitely represents a matrix of complex numbers.

**Comment by eric-wieser** [comment]
I ask because if not, this should be fixable with a `is_commutative = True` class variable?

**Comment by sylee957** [comment]
I think that there are some fundamental flaws of treating commutativity based on its subexpressions.
For example, for Basic, it treats whether an object is commutative or not based on whether its arguments are commutative or not.

https://github.com/sympy/sympy/blob/5d1b2c6f04f583ec0beeea4b8238d4151b497332/sympy/assumptions/handlers/common.py#L47-L51

**Comment by oscarbenjamin** [comment]
I don't understand what is happening because as far as I can tell `e1` and `e2` above are the same but they give different results for `is_commutative`.

**Comment by eric-wieser** [comment]
Note that attempting to fix this by setting `Determinant.is_commutative = True`, results in the error in #19327

**Comment by oscarbenjamin** [comment]
I won't consider this issue fixed until I understand why e1 and e2 give different results.

**Comment by eric-wieser** [comment]
> I won't consider this issue fixed until I understand why e1 and e2 give different results.

A partial explanation:

* `-det_m` is implemented as `Mul._from_args((S.NegativeOne, det_m), det_m.is_commutative)`.
* `-1 * det_m` is implemented as `Mul(S.NegativeOne, det_m)`

**Comment by eric-wieser** [comment]
And here's the thorough explanation:

* `Mul(*args)` calls `AssocOp.__new__`, which splits the arguments into commutative and non-commutative via `Mul.flatten`. The api of `flatten` has no way to indicate "possibly commutative", so the `None` gets turned to `False`
* `Mul._from_args` is happy to propagate `is_commutative=None`

Is `is_commutative == None` even meaningful?

At any rate, it's not correct for `Determinant` - help would be appreciated understanding why making it commutative (#19327) makes sympy crash.

**Comment by oscarbenjamin** [comment]
> Is `is_commutative == None` even meaningful?

No, it's not. Either commutative is true in which case the object is known to commute with anything else or it is not in which case commutative should be False.

**Comment by eric-wieser** [comment]
Note that as of #19354 this is no longer an issue for `Determinant`, but exactly the same problem exists for any custom `Basic` instance with a non-commutative instance in its arguments.

**Comment by brianwa84** [comment]
I wonder if this is could be caused by bool(~True) == True, and [this line](https://github.com/sympy/sympy/blob/master/sympy/assumptions/handlers/common.py#L27) `elif ~Q.commutative(expr) in assumps:` in CommutativePredicate? At least python will soon be outlawing ~bool.
